wgengine/magicsock/magicsock.go

```diff
@@ -2605,66 +2606,82 @@ func (c *Conn) listenPacket(network, host string, port uint16) (net.PacketConn,
+	ruc.mu.Lock()
+	defer ruc.mu.Unlock()
 
+	// Build a list of preferred ports.
+	// Best is the port that the user requested.
+	// Second best is the port that is currently in use.
+	// If those fail, fall back to 0.
+	// The list of preferred ports might contain duplicates.
+	// Slightly wasteful, but no real harm to that.
+	var ports []uint16
 	if c.port != 0 {
-		c.pconn4.mu.Lock()
-		oldPort := c.pconn4.localAddrLocked().Port
-		if err := c.pconn4.pconn.Close(); err != nil {
-			c.logf("magicsock: link change close failed: %v", err)
-		}
-		packetConn, err := c.listenPacket("udp4", host, c.port)
-		if err != nil {
-			c.logf("magicsock: link change unable to bind fixed port %d: %v, falling back to random port", c.port, err)
-			packetConn, err = c.listenPacket("udp4", host, 0)
-			if err != nil {
-				c.logf("magicsock: link change failed to bind random port: %v", err)
-				c.pconn4.mu.Unlock()
-				return
-			}
-			newPort := packetConn.LocalAddr().(*net.UDPAddr).Port
-			c.logf("magicsock: link change rebound port: from %v to %v (failed to get %v)", oldPort, newPort, c.port)
-		} else {
-			c.logf("magicsock: link change rebound port from %d to %d", oldPort, c.port)
+		ports = append(ports, c.port)
+	}
+	if ruc.pconn != nil {
+		curPort := uint16(ruc.localAddrLocked().Port)
+		ports = append(ports, curPort)
+	}
+	ports = append(ports, 0)
+
+	var pconn net.PacketConn
+	var errs []error
+	for _, port := range ports {
+		// Close the existing conn, in case it is sitting on the port we want.
+		err := ruc.closeLocked()
+		if err != nil && !errors.Is(err, net.ErrClosed) && !errors.Is(err, errNilPConn) {
+			return fmt.Errorf("close failed: %w", err)
 		}
-		c.pconn4.pconn = packetConn
-		c.pconn4.mu.Unlock()
-	} else {
-		c.logf("magicsock: link change, binding new port")
-		packetConn, err := c.listenPacket("udp4", host, 0)
+		// Open a new one with the desired port.
+		pconn, err = c.listenPacket(which, host, port)
 		if err != nil {
-			c.logf("magicsock: link change failed to bind new port: %v", err)
-			return
+			err = fmt.Errorf("unable to bind port %d: %w", port, err)
+			errs = append(errs, err)
+			continue
 		}
-		c.pconn4.Reset(packetConn)
+		// Success.
+		ruc.pconn = pconn
+		return nil
+	}
+
+	// Failed to bind, including on port 0 (!).
+	// Set pconn to a dummy conn whose reads block until closed.
+	// This keeps the receive funcs alive for a future in which
+	// we get a link change and we can try binding again.
+	ruc.pconn = newBlockForeverConn()
+	return fmt.Errorf("failed to bind any ports: %w", multierror.New(errs))
+}
+
+// Rebind closes and re-binds the UDP sockets.
+// It should be followed by a call to ReSTUN.
+func (c *Conn) Rebind() {
+	if err := c.bind1(&c.pconn4, "udp4"); err != nil {
+		c.logf("magicsock: Rebind IPv4 failed: %w", err)
 	}
 	c.portMapper.SetLocalPort(c.LocalPort())
+	if err := c.bind1(&c.pconn6, "udp6"); err != nil {
+		c.logf("magicsock: Rebind ignoring IPv6 bind failure: %v", err)
```

Comment:
why's this one "ignoring" but few lines up it also just logs? that's not also ignoring?

Reply:
Thanks, fixed; we now return after logging here.

I also re-read bind1 to see when it returns an error, and decided that failing to close an existing listener shouldn't be fatal, so changed that into additional logging instead. And that made me realize we weren't logging anything when we attempted but failed to bind preferred ports, so I've fixed that too. So bind1 could probably use another look.